### PR TITLE
Add native app authorisation protocol

### PR DESCRIPTION
**Pull request by MichaelDowse**
This is a fork of the original KB-Platform and SongKick gem. This fork has 2 basic motivations;

Add native app authorisation protocol to the standard OAuth2 protocol.
Perform a number of refactoring steps to improve the API and restructure some of the "big classes".
The primary reason for the fork is to add new native app protocols. This is based on draft RFC for native app OAuth, and also uses the Proof Key for Code Exchange scheme recommended for native application that support the "deep link" redirect approach.

As for the refactoring:

The new interface to obtain a parsed and process authorisation looks like this; OAuth2::Provider.parse(user, env).()
The overloaded initialize methods in provider/exchange.rb and provider/authorization.rb have been separated into different concerns.
These classes also have slightly more "dependency injected" validations.
The main token functions have been moved into a lib class SecureCodeScheme

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,2 @@
-source 'https://rubygems.org'
+source 'http://rubygems.org'
```
Comment (MichaelDowse) [inline Gemfile]: Change this back

File: README.rdoc [hunk 2]
```diff
@@ -1,7 +1,4 @@
-= Songkick::OAuth2::Provider
-
-{<img src="https://secure.travis-ci.org/songkick/oauth2-provider.png?branch=master" alt="Build Status" />}[http://travis-ci.org/songkick/oauth2-provider]
-{<img src="https://codeclimate.com/badge.png" />}[https://codeclimate.com/github/songkick/oauth2-provider]
+= OAuth2::Provider
```
Comment (MichaelDowse) [inline README.rdoc:1]: Need to add to our travis

File: LICENSE [hunk 3]
```diff
@@ -1,6 +1,6 @@
 The MIT License
 
-Copyright (c) 2012 Songkick
+Copyright (c) 2016 Flick
```
Comment (MichaelDowse) [inline LICENSE]: Here and below this should be `Flick Electric Co.`

File: README.rdoc [hunk 4]
```diff
@@ -401,7 +398,7 @@ permissions.
 
 == License
 
-Copyright (c) 2010-2012 Songkick.com
+Copyright (c) 2016 FlickElectric.co.nz
```
Comment (MichaelDowse) [inline README.rdoc]: Flick Electric Co.

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+      def self.pkce_string(attributes)
+        cipher = aes_cipher(:encrypt)
+        Base64.urlsafe_encode64(cipher.update(pkce_tokenise(attributes[CODE_CHALLENGE.to_sym], attributes[CODE_CHALLENGE_METHOD.to_sym])) + cipher.final)
+      end
+
+      def self.aes_cipher(direction)
+        cipher = OpenSSL::Cipher::AES.new(256, :CBC)
+        cipher.send(direction)
+        cipher.key = "bff19d8c59f31f68d70e34abae5c93420c17f50bc3c278878593ced6b03d916d"
```
Comment (MichaelDowse) [inline lib/oauth2/lib/secure_code_scheme.rb]: These values probably shouldn't be in the repo

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+      def self.generate_id_token(user)
+        JSON::JWT.new(generate_id_token_hash(user)).sign(rsa_key(ENV[PRIVATE_KEY]), JWT_ALG).to_s
+      end
+
+      def self.rsa_key(key)
+        OpenSSL::PKey::RSA.new(key)
+      end
+
+      def self.generate_id_token_hash(user)
```
Comment (MichaelDowse) [inline lib/oauth2/lib/secure_code_scheme.rb:99]: Not sure this should exist in the library, something very similar already exists at the Identity Service level, can we use that? https://github.com/FlickElectric/identity_service/blob/master/app/models/oauth/token.rb

File: lib/oauth2/model.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    Schema = OAuth2::Schema
+
+    DUPLICATE_RECORD_ERRORS = [
+      /^Mysql::Error:\s+Duplicate\s+entry\b/,
+      /^PG::Error:\s+ERROR:\s+duplicate\s+key\b/,
+      /\bConstraintException\b/
+    ]
+
+    # ActiveRecord::RecordNotUnique was introduced in Rails 3.0 so referring
```
Comment (MichaelDowse) [inline lib/oauth2/model.rb]: Don't bother supporting rails versions < 3

**Comment by MichaelDowse** [comment]
Diff with most of the changes but without the removal of Songkick which makes the diff hard to spot: https://github.com/FlickElectric/oauth2-provider/compare/master...FlickElectric:2a88a3e2695e8b1cb768210eb18ba9cd27df3a02

File: Gemfile (hunk 1 above)
Comment (josephleniston) [inline Gemfile]: Needs to be changed back to `https`.

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        tuple = case attributes[:code_type]
+        when OPAQUE
+          [random_string, :random_string]
+        when PKCE
+          [pkce_string(attributes), :pkce_string]
+        else  # Shouldn't get here, but assume opaque
+          [random_string, :random_string]
+        end
+        id = tuple[0]
+        id = send(tuple[1], attributes) until predicate_function(predicate).call(id)
```
Comment (josephleniston) [inline lib/oauth2/lib/secure_code_scheme.rb]: change to
```ruby
        generation_method = case attributes[:code_type]
        when OPAQUE
          :random_string
        when PKCE
          :pkce_string
        else  # Shouldn't get here, but assume opaque
          :random_string
        end
        id = send(generation_method, attributes) until predicate_function(predicate).call(id)
        id
```

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+      def self.pkce_string(attributes)
+        cipher = aes_cipher(:encrypt)
+        Base64.urlsafe_encode64(cipher.update(pkce_tokenise(attributes[CODE_CHALLENGE.to_sym], attributes[CODE_CHALLENGE_METHOD.to_sym])) + cipher.final)
+      end
+
+      def self.aes_cipher(direction)
+        cipher = OpenSSL::Cipher::AES.new(256, :CBC)
+        cipher.send(direction)
+        cipher.key = "bff19d8c59f31f68d70e34abae5c93420c17f50bc3c278878593ced6b03d916d"
+        cipher.iv = "aa8dbfb30de9bac490cab3aa551376add3462bb9080e0d534f8301cd094f56a7"
```
Comment (josephleniston) [inline lib/oauth2/lib/secure_code_scheme.rb]: These keys should not be embedded they should be in an environment variable.

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      # and produce the original code challenge
+      def self.pkce_run_hash_on_verifier(verifier, method = "S256")
+        Base64.urlsafe_encode64(Digest::SHA256.digest(verifier)).chomp("=")
+      end
+
+
+      # Decrypt the auth code originally sent to the relying party
+      def self.pkce_decrypt(code)
+        cipher = aes_cipher(:decrypt)
+        cipher.update(Base64.urlsafe_decode64(code)) + cipher.final  rescue [nil, nil]
```
Comment (josephleniston) [inline lib/oauth2/lib/secure_code_scheme.rb]: this should rescue a specific exception not any exception.

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      end
+
+
+      def self.random_string(attributes: {})
+        if defined? SecureRandom
+          SecureRandom.hex(TOKEN_SIZE / 8).to_i(16).to_s(36)
```
Comment (josephleniston) [inline lib/oauth2/lib/secure_code_scheme.rb]: This should be required. Why would you want an insecure random?

File: lib/oauth2/lib/secure_code_scheme.rb [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        OpenSSL::PKey::RSA.new(key)
+      end
+
+      def self.generate_id_token_hash(user)
+      {
+        # REQUIRED. Issuer Identifier for the Issuer of the response. The iss
+        # value is a case sensitive URL using the https scheme that contains
+        # scheme, host, and optionally, port number and path components and no
+        # query or fragment components.
+        iss: 'https://id.flick.energy',
```
Comment (josephleniston) [inline lib/oauth2/lib/secure_code_scheme.rb]: This should not be hardcoded. It should be an environment. This is duplicated in identity service so if this going in here then it should be referenced in identity service. This goes for all the values below.

File: lib/oauth2/model/authorization.rb [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+          {code_type: PKCE, code_challenge: attr[CODE_CHALLENGE], code_challenge_method: attr[CODE_CHALLENGE_METHOD]}
+        else
+          {code_type: OPAQUE}
+        end
+      end
+
+      def self.create_access_token
+        Lib::SecureCodeScheme.new.generate(predicate: ->(token) {
+          hash = Lib::SecureCodeScheme.new.hashify(token)
+          Helpers.count(self, :access_token_hash => hash).zero?
```
Comment (josephleniston) [inline lib/oauth2/model/authorization.rb:47]: Why is this different to the test for existence inside `create_refresh_token` method?

File: lib/oauth2/model/client.rb [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      def check_format_of_redirect_uri
+        uri = URI.parse(redirect_uri)
+        errors.add(:redirect_uri, 'must be an absolute URI') unless uri.absolute?
+      rescue
+        errors.add(:redirect_uri, 'must be a URI')
+      end
+
+      def generate_credentials
+        self.client_id = self.class.create_client_id
+        self.client_secret = OAuth2::Lib::SecureCodeScheme.random_string
```
Comment (josephleniston) [inline lib/oauth2/model/client.rb]: You should call `.generate` here to be consistent with other code generation.

File: lib/oauth2/provider.rb [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+      Router.parse(*args)
+    end
+
+    # def self.access_token(*args)
+    #   Router.access_token(*args)
+    # end
+
+    # def self.access_token_from_request(*args)
+    #   Router.access_token_from_request(*args)
+    # end
```
Comment (josephleniston) [inline lib/oauth2/provider.rb]: Remove commented code.

File: lib/oauth2/provider/authorization.rb [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        relying_party ? 302 : 400
+      end
+
+      def valid?
+        @error.nil?
+      end
+
+      def code_predicate_function
+        if native_app_client?
+          ->(code) {true}
```
Comment (josephleniston) [inline lib/oauth2/provider/authorization.rb]: Are you sure this should always return true?

**Comment by josephleniston** [comment]
Can we get some specs for the native app functionality and get the gem running on Travis. Also add some explanation of the new functionality.

**Review by josephleniston — CHANGES_REQUESTED** [review]

File: lib/oauth2/model/authorization.rb (hunk 13 above)
Comment (josephleniston) [reply lib/oauth2/model/authorization.rb:47]: This is original code logic. Looks like refresh_tokens are scoped by client for some reason.

**Pull request merged**